server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java

```diff
@@ -290,30 +290,26 @@ private void deleteConfirmed(GarbageCollectionEnvironment gce,
     cleanUpDeletedTableDirs(gce, candidateMap);
   }
 
-  public void collect(GarbageCollectionEnvironment gce) throws TableNotFoundException, IOException {
-
-    String lastCandidate = "";
-
-    boolean outOfMemory = true;
-    while (outOfMemory) {
-      List<String> candidates = new ArrayList<>();
+  /**
+   * Given a sub-list of possible delection candidates, process and remove valid deletion
+   * candidates.
+   */
+  public void collectBatch(GarbageCollectionEnvironment gce, List<String> currentBatch)
+      throws TableNotFoundException, IOException {
 
-      outOfMemory = getCandidates(gce, lastCandidate, candidates);
-
-      if (candidates.isEmpty())
-        break;
-      else
-        lastCandidate = candidates.get(candidates.size() - 1);
+    long origSize = currentBatch.size();
+    gce.incrementCandidatesStat(origSize);
 
-      long origSize = candidates.size();
-      gce.incrementCandidatesStat(origSize);
+    SortedMap<String,String> candidateMap = makeRelative(currentBatch);
 
-      SortedMap<String,String> candidateMap = makeRelative(candidates);
+    confirmDeletesTrace(gce, candidateMap);
+    gce.incrementInUseStat(origSize - candidateMap.size());
 
-      confirmDeletesTrace(gce, candidateMap);
-      gce.incrementInUseStat(origSize - candidateMap.size());
+    deleteConfirmed(gce, candidateMap);
+  }
```

Comment:
The previous `collect()` method was confusing but it was the only public method of the `GarbageCollectionAlgorithm` so it was at least easier to follow from the one entry point. Since you are refactoring the methods, it might be a good opportunity to clean it up and make it easier to follow. The GC was already confusing but the way you split up the methods, is more confusing to me. It looks like you have the correct business logic between the methods but I think it could be organized better.

It seems the simplest solution would be to keep the same methods but just drop the continue point. Was there a reason you had to create new methods? I saw your comment in the tests about ConcurrentModificationException but I am not seeing where that could happen.

Reply:
I'll take another look, but primarily I was following @keith-turner format suggestion in the inital ticket. Since the call to Ample takes place in the classes implementing the GarbageCollectionEnvironment code, the few short lines seemed to belong in a single method. Not sure if better naming of methods would clear things up. 

Realistically the call to collect could be dropped altogether and just be replaced by a call to processCandidates (since that is all it calls anyway). I just kept it so the existing GC initial call would not be changed.  Then the GCA could have a  public method called' processCandidates' (or maybe even processDeletionCandidates). That seems clearer to me than using  'collect'. Then the the collectBatch call does what is implies. It collects a batch of potential deletion candidates and processes them. Maybe even get away from the 'collect' naming altogether and call it 'processBatch' or something similar.

Reply:
I suspect the ConcurrentModificationException occurs in the test.  I would suggest refactoring the test GCE so that this does not happen rather than the GCA.  The GarbageCollectionAlgorithm was designed to pull the logic of collection into a single place where it could be unit tested.  This change pushes some logic of collection into the GarbCollEnv, resulting in a need to change logic in the test which means the full algorithm is not being tested.

Reply:
I suspect this is what was causing the CME in the unit test https://github.com/apache/accumulo/blob/146b1d3d0dc2ade6f1888c9905d83e37d98aedbe/server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java#L81

Reply:
@keith-turner that was exactly the cause.

Reply:
Still have not looked in depth, so this may not be workable.  Thinking one possible way to avoid the CME would be to make the getCandidates() method in TestGCE return a snapshot.  Something like :

```java
   Iterator<String> getCandidates() {
        return List.copyOf(candidates).iterator();
   }
```